src/app/CommandHandler.h

```diff
@@ -190,9 +226,73 @@ class CommandHandler : public Messaging::ExchangeDelegate
     CHIP_ERROR AddClusterSpecificFailure(const ConcreteCommandPath & aCommandPath, ClusterStatus aClusterStatus);
 
     Protocols::InteractionModel::Status ProcessInvokeRequest(System::PacketBufferHandle && payload, bool isTimedInvoke);
-    CHIP_ERROR PrepareCommand(const ConcreteCommandPath & aCommandPath, bool aStartDataStruct = true);
+
+    /**
+     * This adds a new CommandDataIB element into InvokeResponses for the associated
+     * aRequestCommandPath. This adds up until the `CommandFields` element within
+     * `CommandDataIB`.
+     *
+     * This call will fail if CommandHandler is already in the middle of building a
+     * CommandStatusIB or CommandDataID (ei something has called Perpare*, without
+     * calling Finish*), or is already sending InvokeResponseMessage.
+     *
+     * Upon success, the caller is expected to call `FinishCommand` once they have added
+     * all Data into Fields element of CommandDataIB.
+     *
+     * @param [in] aResponseCommandPath the concrete response path that we are sending to Requester.
+     * @param [in] aPrepareParameters struct containing paramters needs for preparing a command, such as request path.
+     */
+    CHIP_ERROR PrepareInvokeResponseCommand(const ConcreteCommandPath & aResponseCommandPath,
+                                            const InvokeResponseParameters & aPrepareParameters);
+
+    [[deprecated("PrepareCommand now needs the requested command path. Please use PrepareInvokeResponseCommand")]] CHIP_ERROR
+    PrepareCommand(const ConcreteCommandPath & aCommandPath, bool aStartDataStruct = true);
+
+    /**
+     * Finishes the CommandDataIB element within the InvokeResponses.
+     *
+     * Caller must have first successfully called `PrepareInvokeResponseCommand`.
+     *
+     * @param [in] aEndDataStruct end the TLV container for the CommandFields element within
+     *             CommandDataIB.
+     *
+     * @return CHIP_ERROR_INCORRECT_STATE
+     *                      If device has not previously successfully called
+     *                      `PrepareInvokeResponseCommand`.
+     * @retval CHIP_ERROR_BUFFER_TOO_SMALL
+     *                      If writing the values needed to finish the InvokeReponseIB
+     *                      with the current contents of the InvokeResponseMessage
+     *                      would exceed the limit. When this error occurs, it is possible
+     *                      we have already closed some of the IB Builders that were
+     *                      previously started in `PrepareInvokeResponseCommand`.
```

Comment:
Followup: A better API would be if we reserve the space needed to close when the prepare happens, so the "no space" errors happen when trying to write the data there is no space for.